DREAM/Optimization/tsgGradientDescent.hpp

```diff
@@ -0,0 +1,208 @@
+     * \param decrease_coeff Controls how quickly the stepsize is decreased; should be greater than 1
+     * \param num_iterations Number of iterations to perform
+     * \param tolerance Stationarity tolerance; the algorithm terminates early if the stationarity residual computed by
+     *        TasmanianOptimization::compute_stationarity_residual() is less than or equal to \b tolerance
+     * \param state Holds the state of the gradient descent algorithm; see TasOptimization::GradientDescentState
+     */
+    friend OptimizationStatus GradientDescent(const ObjectiveFunctionSingle &f, const GradientFunctionSingle &g,
+                                              const ProjectionFunctionSingle &proj, const double increase_coeff,
+                                              const double decrease_coeff, const int num_iterations, const double tolerance,
+                                              GradientDescentState &state);
+
+    /*! \brief Applies the adaptive NON-PROXIMAL gradient descent algorithm to a gradient descent state (does not require a
+     *         projection function as input).
+     * \ingroup OptimizationAlgorithm Adaptive Non-Proximal Gradient Descent Algorithm
+     *
+     * Runs \b num_iterations of the adaptive gradient descent algorithm to a gradient descent \b state to minimize the function
+     * \b f over an unconstrained domain.  This method is guaranteed to converge to a stationary point if the gradient of \b f is
+     * Lipschitz continuous everywhere. Requires two line search coefficients. Outputs a TasmanianOptimization::OptimizationStatus
+     * structure that contains information about the last iterate.
+     *
+     * \param f Objective function to be minimized
+     * \param g Gradient of the objective function
+     * \param increase_coeff Controls how quickly the stepsize is increased; should be greater than 1
+     * \param decrease_coeff Controls how quickly the stepsize is decreased; should be greater than 1
+     * \param num_iterations Number of iterations to perform
+     * \param tolerance Stationarity tolerance; the algorithm terminates early if the stationarity residual computed by
+     *        TasmanianOptimization::compute_stationarity_residual() is less than or equal to \b tolerance
+     * \param state Holds the state of the gradient descent algorithm, see TasOptimization::GradientDescentState
+     */
+    friend OptimizationStatus GradientDescent(const ObjectiveFunctionSingle &f, const GradientFunctionSingle &g,
+                                              const double increase_coeff, const double decrease_coeff, const int num_iterations,
+                                              const double tolerance, GradientDescentState &state);
+
+  protected:
+    #ifndef __TASMANIAN_DOXYGEN_SKIP_INTERNAL
+    //! \brief Returns a reference to the current point.
+    inline std::vector<double> &getXRef() {return x;}
```

Comment:
Changed. The signatures in the various `GradientDescent` functions all have `GradientDescentState &state` instead of `std::vector<double> &state`, and I have removed the `GradientDescentState::getXRef()` method.